genrl/classical/bandit/policies.py

```diff
@@ -0,0 +1,791 @@
+
+        :returns: List of rewards
+        :rtype: list
+        """
+        return self._reward_hist
+
+    @property
+    def counts(self):
+        """
+        Get the number of times each action has been taken
+
+        :returns: Numpy array with count for each action
+        :rtype: numpy.ndarray
+        """
+        return self._counts
+
+    def select_action(self, t):
+        """
+        Select an action
+
+        This method needs to be implemented in the specific policy.
+
+        :param t: timestep to choose action for
+        :type t: int
+        :returns: Selected action
+        :rtype: int
+        """
+        raise NotImplementedError
+
+    def update_params(self, action, reward):
+        """
+        Update parmeters for the policy
+
+        This method needs to be implemented in the specific policy.
+
+        :param action: action taken for the step
+        :param reward: reward obtained for the step
+        :type action: int
+        :type reward: float
+        """
+        raise NotImplementedError
+
+    def learn(self, n_timesteps=None):
+        """
+        Learn to solve the environment over given number of timesteps
+
+        This method needs to be implemented in the specific policy.
+
+        :param n_timesteps: number of steps to learn for
+        :type: int 
+        """
+        raise NotImplementedError
+
+
+class EpsGreedyPolicy(BanditPolicy):
+    """
+    Multi-Armed Bandit Solver with Epsilon Greedy Action Selection Strategy.
+
+    Refer to Section 2.3 of Reinforcement Learning: An Introduction.
+
+    :param bandit: The Bandit to solve
+    :param eps: Probability with which a random action is to be selected.
+    :type bandit: Bandit type object 
+    :type eps: float 
+    """
+
+    def __init__(self, bandit, eps=0.05):
+        super(EpsGreedyPolicy, self).__init__(bandit)
+        self._eps = eps
+        self._Q = np.zeros(bandit.arms)
+
+    @property
+    def eps(self):
+        """
+        Get the asscoiated epsilon for the policy 
+
+        :returns: Probability with which a random action is to be selected
+        :rtype: float
+        """
+        return self._eps
+
+    @property
+    def Q(self):
+        """
+        Get the q values assigned by the policy to all actions
+
+        :returns: Numpy array of q values for all actions
+        :rtype: numpy.ndarray
+        """
+        return self._Q
+
+    def learn(self, n_timesteps=1000):
+        """
+        Learn to solve the environment over given number of timesteps
+
+        Selects action, takes a step in the bandit and then updates
+        the parameters according to the reward received.
+
+        :param n_timesteps: number of steps to learn for
+        :type: int 
+        """
+        for t in range(n_timesteps):
+            action = self.select_action(t)
+            reward = self._bandit.step(action)
+            self.update_params(action, reward)
+            self.reward_hist.append(reward)
```

Comment:
The `learn` function is almost the same in every class. Can you find a way to restrict to the base classes only?

Reply:
> The `learn` function is almost the same in every class. Can you find a way to restrict to the base classes only?

For DCB, we would have slightly different learn function since we would have observations too right?

Should I modify the API so that it fit those requirements?  I.e. `select_action` would have another paprameter for observation and `learn` would have to pass observation to `select_action`